draft-davidben-tls-merkle-tree-certs.md

```diff
@@ -349,12 +349,12 @@ struct {
 
 struct {
     SubjectType subject_type;
-    opaque subject_info<0..2^24-1>;
+    opaque subject_info_hash[hash.length];
```

Comment:
I suppose, under this model (if we want to keep it!), we'd have the `Assertion` struct which is the general, unhashed one. But then the tree is computed over:

```
struct {
    SubjectType subject_type;
    HashValueSHA256 subject_type_hash;
    Claim claims<0..2^24-1>;
} MerkleTreeAssertionSHA256;
```

Reply:
I like pulling out `Assertion` like this in the ACME flow.

(Minor point: if I understand correctly, if we'd want to leave out the hash in the BikeshedCertificate, we'd have three different assertion structs: the one with the public key for the "CSR"; the one with the hash for the tree and CA to publish; and the one without the hash for the BikeshedCertificate.)

Originally I thought we could group almost everything we propose under MTC, except perhaps for the negotiation mechanism itself. But now it came into focus for me that you intend most of the structures to be reused by instances different from MTC. Perhaps we should name this more general system, and move the hash into it. Perhaps: `web-sha256` and have MTC be a certificate type of it `web-sha256/mtc`.

Reply:
Does the CSR one need the hash? I figured the hash would just be private to MTCs.

Reply:
(It's a separate question whether proof-of-possession is actually required. I'm not sure either way.)

Reply:
Oh sure I meant that they would have the unhashed version. I.e. generic is unhashed and both CSRs and certs use unhashed. Hashed only shows up as an implementation detail of building the tree.

Reply:
And what the MTC CAs serve.

Reply:
Ah yeah that too. I always forget about that one, I think because we never actually defined the darn HTTP interface. :-) Anyway, my point is more that we only need two structures, not one. I think the breakdown goes:

* Assertion is the one as-is. MerkleTreeAssertionSHA256 is the one with only the hash. You can compute it from Assertion by just hashing the subject.
* `GET {prefix}/batch/{number}/assertions` returns a list of MerkleTreeAssertionSHA256
* HashAssertion and the tree building operation takes MerkleTreeAssertionSHA256 as input
* Everything else (BikeshedCertificate, issuance process, etc.) uses Assertion